### PR TITLE
initial commit of example code

### PR DESCRIPTION
**Pull request by jasoncoposky**

### CODE REVIEW

File: development/plugins/authentication_api.cmake [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    )
+
+target_link_libraries(
+    ${TARGET_NAME}
+    PRIVATE
+    irods_common
+    irods_server
+    irods_plugin_dependencies
+    ${IRODS_EXTERNALS_FULLPATH_BOOST}/lib/libboost_filesystem.so
+    ${IRODS_EXTERNALS_FULLPATH_BOOST}/lib/libboost_filesystem.so
```
Comment (trel) [inline development/plugins/authentication_api.cmake:40]: Duplicate

File: development/plugins/libirods-api_plugin-authentication.cpp [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+        return api->call_handler<bytesBuf_t*, bytesBuf_t**>(comm, inp, out);
+    }
+
+#ifdef RODS_SERVER
+    #define CALL_AUTHENTICATE call_authenticate
+#else
+    #define CALL_AUTHENTICATE NULL
+#endif
+
+    static const uint32_t AUTHENTICATION_APN{110000};
```
Comment (trel) [inline development/plugins/libirods-api_plugin-authentication.cpp:64]: Does this number need to be coordinated anywhere?    How was it selected?

File: development/plugins/authentication_plugin_framework.hpp [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+    } // throw_if_missing
+
+
+
+    void throw_if_empty(const std::string& n, const std::string& p)
+    {
+        if(p.empty()) {
+                THROW(
+                    SYS_INVALID_INPUT_PARAM,
+                    boost::format("paramerer [%s] is empty for authentication flow")
```
Comment (trel) [inline development/plugins/authentication_plugin_framework.hpp:125]: ”paramerer”

File: development/plugins/authentication_plugin.cmake [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+    target_link_libraries(
+        ${IRODS_CURRENT_PLUGIN_TARGET}
+        PRIVATE
+        irods_common
+        irods_server
+        irods_plugin_dependencies
+        irods_client
+        crypto
+        ${IRODS_EXTERNALS_FULLPATH_BOOST}/lib/libboost_filesystem.so
+        ${IRODS_EXTERNALS_FULLPATH_BOOST}/lib/libboost_filesystem.so
```
Comment (trel) [inline development/plugins/authentication_plugin.cmake:50]: Duplicate

File: development/plugins/libirods-api_plugin-authentication.cpp (hunk 2 above)
Comment (jasoncoposky) [reply development/plugins/libirods-api_plugin-authentication.cpp:64]: it is impossible to coordinate api numbers, so i picked an arbitrarily large number.

File: development/plugins/authentication_api.cmake [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+  COMPONENT ${IRODS_PACKAGE_COMPONENT_POLICY_NAME}
+  )
+
+#install(
+#  FILES
+#  packaging/test_plugin_policy_engine-${POLICY_NAME}.py
+#  DESTINATION var/lib/irods/scripts/irods/test
+#  COMPONENT ${IRODS_PACKAGE_COMPONENT_POLICY_NAME}
+#  )
+
```
Comment (trel) [inline development/plugins/authentication_api.cmake:62]: torch?

File: development/plugins/authentication_plugin.cmake [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+      )
+endforeach()
+
+#install(
+#  FILES
+#  packaging/test_plugin_policy_engine-${POLICY_NAME}.py
+#  DESTINATION var/lib/irods/scripts/irods/test
+#  COMPONENT ${IRODS_PACKAGE_COMPONENT_POLICY_NAME}
+#  )
+
```
Comment (trel) [inline development/plugins/authentication_plugin.cmake:73]: torch?

File: development/plugins/authentication_plugin_framework.hpp [hunk 7]
```diff
@@ -0,0 +1,258 @@
...
+
+    void throw_if_empty(const std::string& n, const std::string& p)
+    {
+        if(p.empty()) {
+                THROW(
+                    SYS_INVALID_INPUT_PARAM,
+                    boost::format("parameter [%s] is empty for authentication flow")
+                    % n);
+        }
+    } // throw_if_missing
```
Comment (trel) [inline development/plugins/authentication_plugin_framework.hpp:128]: throw_if_empty

File: development/plugins/authentication_plugin_framework.hpp [hunk 8]
```diff
@@ -0,0 +1,258 @@
...
+        auto auth = resolve_authentication_plugin(scheme, "client");
+
+        std::string next_operation{irods::AUTH_CLIENT_START};
+
+        json req{}, resp{};
+
+        req["scheme"] = scheme;
+        req["next_operation"] = next_operation;
+
+        while(true) { // TODO: neeed additional exit critiera?
```
Comment (trel) [inline development/plugins/authentication_plugin_framework.hpp:238]: spelling - need, criteria

File: development/plugins/libirods-authentication_plugin-native.cpp [hunk 9]
```diff
@@ -0,0 +1,450 @@
...
+                        }
+                    }
+                }
+                else if ( 0 == strcmp(comm->proxyUser.userName,  comm->clientUser.userName ) ) {
+                    authCheckOut->clientPrivLevel = authCheckOut->privLevel;
+                }
+
+                auto ret = check_proxy_user_privileges(comm, authCheckOut->privLevel );
+                if ( !ret.ok() ) {
+                    THROW(ret.code(), "Check proxy user priviledges failed.");
```
Comment (trel) [inline development/plugins/libirods-authentication_plugin-native.cpp:390]: typo - privileges

File: development/plugins/libirods-authentication_plugin-native.cpp [hunk 10]
```diff
@@ -0,0 +1,450 @@
...
+        } // native_auth_agent_verify
+
+
+        // =-=-=-=-=-=-=-
+        // stub for ops that the native plug does
+        // not need to support
+        json native_auth_agent_start(rsComm_t*, const json&)
+        {
+            return {};
+        } // native_auth_success_stub
```
Comment (trel) [inline development/plugins/libirods-authentication_plugin-native.cpp:440]: native_auth_agent_start

**Pull request merged**